### PR TITLE
Owise encoding decision.

### PR DESCRIPTION
**Pull request by virgil-serbanuta**
###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`
- [ ] Style conformance: `stylish-haskell`

---

### CODE REVIEW

**Comment by dwightguth** [comment]
This looks reasonable to me but I'm having a little trouble following the details. Can you please clarify for me what, if anything, needs to be changed in the k to kore translation as a result of this decision?

**Comment by traiansf** [comment]
@dwightguth the generation of the first argument of the `\implies` for the `owise` rule will need to change.  

Say the `owise` rule is `f(ot1, .., otn) => ot requires or ensures oe`, 
and assume non-owise rules (let's use just one for exemplification) are of the form `f(t1, .., tn) => t requires r ensures e` with variables `X`

Currently, the precondition is formed as an `\and` between `or`, the requires of the `owise` rule, and the negation of a disjunction over `\ceil(\exists{X}(\and(r, f(t1,..,tn)))) for all non-owise rules.

We would like the disjunction items to change and become of the form:
`\exists{X}(\and(r, \and(\ceil(\and(t1, ot1)), \and(\ceil(\and(t2, ot2)), ...,\and(\ceil(\and(tn,otn), \top()) ...)))`

 This way it would be clear that the `owise` substitution that matches the term to be rewritten is used to unify the subterms of the function from the non-owise rules.

File: design-decisions/2018-09-03-Owise-encoding.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+                        And{p = 1, k}(
+                            Ceil(And(ti_p(xi1, ..., xi_ji), tp(x1, ..., xj)))
+                        )
+                        Pi(xi1, ..., xi_ji)
+                    )
+                )
+            )
+        )
+    )
+```
```
Comment (hreada) [inline design-decisions/2018-09-03-Owise-encoding.md]: I'm in favor of the encoding. It's correct. It works even for cases when `t_1,...,t_k` are not constructors, but any symbols.

File: design-decisions/2018-09-03-Owise-encoding.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        And(ti_p(xi1, ..., xi_ji), tp(x1, ..., xj))
+    )
+)
+```
+
+Equals vs And explanation
+-------------------------
+
+Let us say that we have a function `f` of one argument and a non-functional
+constructor `C` also of one argument. Let us say that our function is defined
```
[the remark this answers is not in the document]
Comment (virgil-serbanuta) [reply design-decisions/2018-09-03-Owise-encoding.md]: I think that in practice they are. However, the semantics-of-k document, page 6, allows non-functional constructors.

I think that none of those axioms implies functionality, although I didn't think through all the implications of the "No confusion, same constructor" axiom. Also see the following quote from the same page:

"Additionally, if each ci is functional, then we call them functional constructors."

[the remark this answers is not in the document]
Comment (u--) [reply design-decisions/2018-09-03-Owise-encoding.md]: IMO, that should be added. I can't think of a single case where non-functional constructors make sense. But that's a different PR.

**Pull request merged**